iroh-willow/src/store/auth.rs

```diff
@@ -0,0 +1,295 @@
+//! Authentication backend for Willow.
+//!
+//! Manages capabilities.
+
+use std::collections::{HashMap, HashSet};
+
+use anyhow::Result;
+use ed25519_dalek::SignatureError;
+use meadowcap::{IsCommunal, NamespaceIsNotCommunalError, OwnedCapabilityCreationError};
+use tracing::{debug, trace};
+
+use crate::{
+    interest::{
+        AreaOfInterestSelector, CapSelector, CapabilityPack, DelegateTo, InterestMap, Interests,
+        InvalidCapabilityPack, RestrictArea,
+    },
+    proto::{
+        data_model::WriteCapability,
+        grouping::AreaOfInterest,
+        keys::{NamespaceId, UserId},
+        meadowcap::{
+            AccessMode, FailedDelegationError, McCapability, McSubspaceCapability,
+            ReadAuthorisation,
+        },
+    },
+    store::traits::{CapsStorage, SecretStorage, SecretStoreError, Storage},
+};
+
+#[derive(Debug, Clone)]
+pub struct Auth<S: Storage> {
+    secrets: S::Secrets,
+    caps: S::Caps,
+}
+
+impl<S: Storage> Auth<S> {
+    pub fn new(secrets: S::Secrets, caps: S::Caps) -> Self {
+        Self { secrets, caps }
+    }
+    pub fn get_write_cap(
+        &self,
+        selector: &CapSelector,
+    ) -> Result<Option<WriteCapability>, AuthError> {
+        let cap = self.caps.get_write_cap(selector)?;
+        Ok(cap)
+    }
+
+    pub fn get_read_cap(
```

Comment:
It would be nice to have a unique ID for the capability to query it exactly by the ID, because a capability selector could return more than one capability. Also, when we want to list capabilities (e.g. in some admin UI) it would be pleasant to have some readable names associated with the capabilities IDs (as I mentioned in a comment about namespace store ergonomics here https://github.com/n0-computer/iroh/pull/2231#pullrequestreview-2366756832).